staging/kos/pkg/controllers/subnet/validator.go

```diff
@@ -0,0 +1,458 @@
+			conflictsMsgs = append(conflictsMsgs, fmt.Sprintf("CIDR overlaps with %s's (%s)", potentialRival.NamespacedName, pr.Spec.IPv4))
+		}
+		if potentialRival.NSConflict(candidate) {
+			glog.V(2).Infof("Namespace conflict found between %s and %s.", candidate.NamespacedName, potentialRival.NamespacedName)
+			conflictsMsgs = append(conflictsMsgs, fmt.Sprintf("same VNI but different namespace wrt %s", potentialRival.NamespacedName))
+		}
+
+		// Record the conflict in the conflicts cache.
+		if err := v.recordConflict(potentialRival, candidate); err != nil {
+			return conflictsMsgs, conflictFound, err
+		}
+	}
+
+	return conflictsMsgs, conflictFound, nil
+}
+
+func (v *Validator) updateSubnetValidity(s *netv1a1.Subnet, validated bool, validationErrors []string) error {
+	sCopy := s.DeepCopy()
+
+	sCopy.Status.Validated = validated
+	sCopy.Status.Errors.Validation = validationErrors
+
+	_, err := v.netIfc.Subnets(sCopy.Namespace).Update(sCopy)
+	switch {
+	case err == nil:
+		nsn := k8stypes.NamespacedName{
+			Namespace: s.Namespace,
+			Name:      s.Name,
+		}
+		v.updateStaleRV(nsn, s.ResourceVersion)
+	case doNotRetryUpdate(err):
+		glog.Errorf("failed to update subnet from %#+v to %#+v: %s. There will be no retry because of the nature of the error", s, sCopy, err.Error())
+	case !doNotRetryUpdate(err):
+		return fmt.Errorf("failed to update subnet from %#+v to %#+v: %s", s, sCopy, err.Error())
+	}
+
+	return nil
+}
+
+func (v *Validator) recordConflict(enroller, enrollee *subnet.Summary) error {
+	v.conflictsMutex.Lock()
+	defer v.conflictsMutex.Unlock()
+
+	c := v.conflicts[enroller.NamespacedName]
+
+	if c == nil {
+		return fmt.Errorf("registration of %s as a rival of %s failed: %s's conflicts cache not found", enrollee.NamespacedName, enroller.NamespacedName, enroller.NamespacedName)
+	}
+
+	if !enroller.Equal(c.ownerSummary) {
+		// If we're here the version of the enroller recorded in its conflicts
+		// cache does not match the version of the enroller we got with the live
+		// list to the API server: one of the two is stale. Return an error so
+		// that the caller can wait a little bit and retry (hopefully the
+		// version skew has resolved by then).
+		return fmt.Errorf("registration of %s as a rival of %s failed: mismatch between %s's live data (%#+v) and conflicts cache data (%#+v)", enrollee.NamespacedName, enroller.NamespacedName, enroller.NamespacedName, enroller, c.ownerSummary)
+	}
+
+	c.rivals = append(c.rivals, enrollee.NamespacedName)
+	return nil
+}
+
+func (v *Validator) updateStaleRV(s k8stypes.NamespacedName, rv string) {
+	v.staleRVsMutex.Lock()
+	defer v.staleRVsMutex.Unlock()
+
+	v.staleRVs[s] = rv
+}
+
+func doNotRetryUpdate(e error) bool {
+	return k8serrors.IsUnauthorized(e) ||
+		k8serrors.IsBadRequest(e) ||
+		k8serrors.IsForbidden(e) ||
+		k8serrors.IsNotAcceptable(e) ||
+		k8serrors.IsUnsupportedMediaType(e) ||
+		k8serrors.IsMethodNotSupported(e) ||
+		k8serrors.IsInvalid(e) ||
+		k8serrors.IsGone(e)
```

Comment:
`k8serrors.IsGone(e)` will never be true in the usage here.  The "gone" condition is only returned in one case: asking for a watch from a ResourceVersion that is too old.

It seems a bit verbose to have two "doNotRetry" methods that are very similar.  I suggesting unifying the two, naming it "requestMalformed", and making it test `IsInvalid` but not `IsGone`.

Reply:
Fixed in 8ac9883.